pkg/config/reloader.go

```diff
@@ -0,0 +1,164 @@
+	logger log.Logger,
+	reg prometheus.Registerer,
+	filename string,
+	reloaders []ComponentReloader,
+) (*ConfigReloader, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to establish config watcher", "err", err)
+		return nil, err
+	}
+
+	if err := watcher.Add(filename); err != nil {
+		level.Error(logger).Log("msg", "failed to start watching config file", "err", err, "path", filename)
+		return nil, err
+	}
+
+	r := &ConfigReloader{
+		logger:   logger,
+		filename: filename,
+		watcher:  watcher,
+
+		reloaders: reloaders,
+
+		triggerReload: make(chan struct{}, 1),
+
+		configSuccess: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "parca_config_last_reload_successful",
+			Help: "Whether the last configuration reload attempt was successful.",
+		}),
+		configSuccessTime: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "parca_config_last_reload_success_timestamp_seconds",
+			Help: "Timestamp of the last successful configuration reload.",
+		}),
+	}
+
+	reg.MustRegister(
+		r.configSuccess,
+		r.configSuccessTime,
+	)
+
+	return r, nil
+}
+
+func (r *ConfigReloader) watchFile() {
+	for {
+		select {
+		case event, ok := <-r.watcher.Events:
+			if !ok {
+				return
+			}
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				level.Debug(r.logger).Log("msg", "config file has been modified")
+				r.triggerReload <- struct{}{}
+			}
+		case err, ok := <-r.watcher.Errors:
+			if !ok {
+				return
+			}
+			level.Error(r.logger).Log("msg", "error encountered while watching config file", "err", err)
+		}
+	}
+}
+
+func (r *ConfigReloader) reloadFile() (err error) {
+	start := time.Now()
+	timings := []interface{}{}
+	level.Info(r.logger).Log("msg", "loading configuration file", "filename", r.filename)
+
+	defer func() {
+		if err == nil {
+			r.configSuccess.Set(1)
+			r.configSuccessTime.SetToCurrentTime()
+		} else {
+			r.configSuccess.Set(0)
+		}
+	}()
+
+	cfg, err := LoadFile(r.filename)
+	if err != nil {
+		return fmt.Errorf("failed to load configuration: %w", err)
+	}
+
+	if err = cfg.Validate(); err != nil {
+		return fmt.Errorf("parsed configuration invalid (--config-path=%q): %w", r.filename, err)
+	}
+
+	failed := false
+	for _, rl := range r.reloaders {
+		rstart := time.Now()
+		if err := rl.Reloader(cfg); err != nil {
+			level.Error(r.logger).Log("msg", "failed to apply configuration", "err", err)
+			failed = true
+		}
+		timings = append(timings, rl.Name, time.Since(rstart))
+	}
+	if failed {
+		return fmt.Errorf("one or more errors occurred while applying the new configuration (--config-path=%q)", r.filename)
+	}
+
+	l := []interface{}{"msg", "completed loading of configuration file", "filename", r.filename, "totalDuration", time.Since(start)}
+	level.Info(r.logger).Log(append(l, timings...)...)
+	return nil
+}
+
+// Run starts watching the config file and wait for reload triggers.
+func (r *ConfigReloader) Run(ctx context.Context) error {
+	go r.watchFile()
```

Comment:
It probably makes sense to use the `run.Group` we use here too. 
Not sure we will notice if this goroutine stops right now.

Reply:
It seems it would only be useful if the goroutine can return an error, but here it only exits if one of the watcher's channels is closed, it happens when `watcher.Close()` is called, which is the normal behavior when Parca exits. I have added debug logs to troubleshoot potential issues around the channel closing if ever it arises, please let me know if it is enough

fyi (and not that it justifies the handling of `watchFile()`) there is this goroutine which is in the same case:
https://github.com/parca-dev/parca/blob/1c12bfb5ca4009c45f61c0aa711c3b5b8a4a6e7c/pkg/scrape/manager.go#L157-L160

If I understand correctly, it triggers config reload for the service discovery, reloads which used to panic as I pointed above, probably nobody had a case where discovery would change a target's config

Reply:
Ok. Let's try with the logs for now. :+1:

Reply:
I think at the very least we should pass the context to the `watchFile` method and check for the cancellation.

Reply:
It happens in cascade from the parent goroutine, do you still think it necessary to pass it directly in addition to the channels closure?

```go
		case <-ctx.Done():
			r.watcher.Close()
			return nil
		}
```

https://github.com/parca-dev/parca/pull/1141/files#diff-c1a28a96ac2f667229d0cfd4d99f4b4106c61ce7626adeedbd4a4c03f54e09e3R161-R164

Reply:
Aha, I have missed this part! Then it should be all good.
Intuitively I expect to see the context passed down and handled in the goroutine itself.